File: packages/core/src/assertion.ts
import {
  EmbeddedJWK,
  SignJWT,
  calculateJwkThumbprint,
  decodeProtectedHeader,
  jwtVerify,
  type JWK,
} from "jose";
import { createHash, randomBytes } from "node:crypto";
import { ErrorCodes, ToolgateError } from "./errors.js";
import { importSigningKey, importVerifyKey } from "./keys.js";

export const CLIENT_ASSERTION_TYP = "tg-client+jwt";
export const POP_PROOF_TYP = "tg-pop+jwt";
const PROOF_MAX_AGE_SECONDS = 60;

// ---------------------------------------------------------------------------
// Client assertion (RFC 7523 private_key_jwt style): how an agent
// authenticates to the control plane token endpoint. No shared secrets.
// ---------------------------------------------------------------------------

export async function signClientAssertion(
  agentPrivateJwk: JWK,
  options: { agentId: string; tokenUrl: string; ttlSeconds?: number },
): Promise<string> {
  const key = await importSigningKey(agentPrivateJwk);
  return new SignJWT({})
    .setProtectedHeader({ alg: "EdDSA", typ: CLIENT_ASSERTION_TYP })
    .setIssuer(options.agentId)
    .setSubject(options.agentId)
    .setAudience(options.tokenUrl)
    .setIssuedAt()
    .setExpirationTime(new Date(Date.now() + (options.ttlSeconds ?? 60) * 1000))
    .setJti(randomBytes(12).toString("base64url"))
    .sign(key);
}

export async function verifyClientAssertion(
  agentPublicJwk: JWK,
  assertion: string,
  options: { expectedAudience: string },
): Promise<{ agentId: string; jti: string }> {
  const key = await importVerifyKey(agentPublicJwk);
  try {
    const { payload } = await jwtVerify(assertion, key, {
      audience: options.expectedAudience,
      typ: CLIENT_ASSERTION_TYP,
    });
    if (!payload.iss || payload.iss !== payload.sub || typeof payload.jti !== "string") {
      throw new Error("client assertion must have iss === sub and a jti");
    }
    return { agentId: payload.sub as string, jti: payload.jti };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new ToolgateError(ErrorCodes.TOKEN_INVALID, `client assertion rejected: ${message}`);
  }
}

// ---------------------------------------------------------------------------
// Proof of possession (DPoP-style): every gate call carries a one-time proof
// signed by the agent key named in the capability token's cnf.jkt. A stolen
// token is useless without the key.
// ---------------------------------------------------------------------------

export function accessTokenHash(token: string): string {
  return createHash("sha256").update(token).digest("base64url");
}

export async function signPopProof(
  agentPrivateJwk: JWK,
  options: { htm: string; htu: string; accessToken: string },
): Promise<string> {
  const key = await importSigningKey(agentPrivateJwk);
  const { kid: _kid, alg: _alg, ...publicJwk } = publicJwkFromPrivate(agentPrivateJwk);
  return new SignJWT({
    htm: options.htm.toUpperCase(),
    htu: options.htu,
    ath: accessTokenHash(options.accessToken),
  })
    .setProtectedHeader({ alg: "EdDSA", typ: POP_PROOF_TYP, jwk: publicJwk })
    .setIssuedAt()
    .setJti(randomBytes(12).toString("base64url"))
    .sign(key);
}

export interface VerifiedPopProof {
  jti: string;
  jkt: string;
}

export async function verifyPopProof(
  proof: string,
  options: { expectedJkt: string; htm: string; htu: string; accessToken: string },
): Promise<VerifiedPopProof> {
  try {
    const header = decodeProtectedHeader(proof);
    if (header.typ !== POP_PROOF_TYP || !header.jwk) {
      throw new Error("missing typ or embedded jwk");
    }
    const jkt = await calculateJwkThumbprint(header.jwk as JWK);
    if (jkt !== options.expectedJkt) {
      throw new Error("proof key does not match token cnf.jkt");
    }
    const { payload } = await jwtVerify(proof, EmbeddedJWK, {
      typ: POP_PROOF_TYP,
      maxTokenAge: PROOF_MAX_AGE_SECONDS,
      clockTolerance: 5,
    });
    if (payload.htm !== options.htm.toUpperCase()) throw new Error("htm mismatch");
    if (payload.htu !== options.htu) throw new Error("htu mismatch");
    if (payload.ath !== accessTokenHash(options.accessToken)) throw new Error("ath mismatch");
    if (typeof payload.jti !== "string") throw new Error("missing jti");
    return { jti: payload.jti, jkt };
  } catch (err) {
    if (err instanceof ToolgateError) throw err;
    const message = err instanceof Error ? err.message : String(err);
    throw new ToolgateError(ErrorCodes.PROOF_INVALID, `proof-of-possession rejected: ${message}`);
  }
}

/** Ed25519 public JWK is the private JWK minus the secret scalar `d`. */
function publicJwkFromPrivate(privateJwk: JWK): JWK {
  const { d: _d, ...rest } = privateJwk;
  return rest;
}

File: src/toolgate/core/assertion.py
"""Client assertions (RFC 7523 style) and DPoP-style proof-of-possession.

The PoP layer is hand-implemented: as of 2026 no maintained Python JOSE
library ships DPoP (authlib #315 open since 2021). The proof format is
Toolgate's own (`tg-pop+jwt`), matching RFC 9449's model: one-time JWS signed
by the agent key, embedding the public JWK, bound to method + URL + token hash.
"""

import base64
import hashlib
import json
import secrets
import time
from dataclasses import dataclass
from typing import Any

from jwcrypto import jwk, jws, jwt

from .errors import ErrorCodes, ToolgateError
from .keys import KeyLike, public_jwk_from_private, to_jwk

CLIENT_ASSERTION_TYP = "tg-client+jwt"
POP_PROOF_TYP = "tg-pop+jwt"
_PROOF_MAX_AGE_SECONDS = 60
_PROOF_CLOCK_SKEW_SECONDS = 5

# ---------------------------------------------------------------------------
# Client assertion: how an agent authenticates to the control plane token
# endpoint. No shared secrets.
# ---------------------------------------------------------------------------


def sign_client_assertion(
    agent_private_jwk: KeyLike,
    *,
    agent_id: str,
    token_url: str,
    ttl_seconds: int = 60,
) -> str:
    key = to_jwk(agent_private_jwk)
    now = int(time.time())
    token = jwt.JWT(
        header={"alg": "EdDSA", "typ": CLIENT_ASSERTION_TYP},
        claims={
            "iss": agent_id,
            "sub": agent_id,
            "aud": token_url,
            "iat": now,
            "exp": now + ttl_seconds,
            "jti": secrets.token_urlsafe(12),
        },
    )
    token.make_signed_token(key)
    return token.serialize()


@dataclass(frozen=True)
class VerifiedAssertion:
    agent_id: str
    jti: str


def verify_client_assertion(
    agent_public_jwk: KeyLike,
    assertion: str,
    *,
    expected_audience: str,
) -> VerifiedAssertion:
    key = to_jwk(agent_public_jwk)
    verifier = jwt.JWT(check_claims={"aud": expected_audience, "exp": None})
    verifier.leeway = _PROOF_CLOCK_SKEW_SECONDS
    try:
        verifier.deserialize(assertion, key)
        header = json.loads(verifier.header)
        if header.get("typ") != CLIENT_ASSERTION_TYP:
            raise ValueError("client assertion has wrong typ")
        claims = json.loads(verifier.claims)
        if not claims.get("iss") or claims["iss"] != claims.get("sub"):
            raise ValueError("client assertion must have iss == sub")
        if not isinstance(claims.get("jti"), str):
            raise ValueError("client assertion must have a jti")
        return VerifiedAssertion(agent_id=claims["sub"], jti=claims["jti"])
    except ToolgateError:
        raise
    except Exception as err:
        raise ToolgateError(
            ErrorCodes.TOKEN_INVALID, f"client assertion rejected: {err}"
        ) from err


# ---------------------------------------------------------------------------
# Proof of possession: every gate call carries a one-time proof signed by the
# agent key named in the capability token's cnf.jkt. A stolen token is useless
# without the key.
# ---------------------------------------------------------------------------


def access_token_hash(token: str) -> str:
    digest = hashlib.sha256(token.encode()).digest()
    return base64.urlsafe_b64encode(digest).rstrip(b"=").decode()


def sign_pop_proof(
    agent_private_jwk: KeyLike,
    *,
    htm: str,
    htu: str,
    access_token: str,
) -> str:
    key = to_jwk(agent_private_jwk)
    if isinstance(agent_private_jwk, dict):
        header_jwk = public_jwk_from_private(agent_private_jwk)
    else:
        exported: dict[str, Any] = json.loads(key.export_public())
        header_jwk = {k: exported[k] for k in ("kty", "crv", "x")}
    token = jwt.JWT(
        header={
            "alg": "EdDSA",
            "typ": POP_PROOF_TYP,
            "jwk": header_jwk,
        },
        claims={
            "htm": htm.upper(),
            "htu": htu,
            "ath": access_token_hash(access_token),
            "iat": int(time.time()),
            "jti": secrets.token_urlsafe(12),
        },
    )
    token.make_signed_token(key)
    return token.serialize()


@dataclass(frozen=True)
class VerifiedPopProof:
    jti: str
    jkt: str


def verify_pop_proof(
    proof: str,
    *,
    expected_jkt: str,
    htm: str,
    htu: str,
    access_token: str,
) -> VerifiedPopProof:
    try:
        signature = jws.JWS()
        signature.deserialize(proof)
        header = signature.jose_header
        if header.get("typ") != POP_PROOF_TYP or "jwk" not in header:
            raise ValueError("missing typ or embedded jwk")

        embedded = jwk.JWK(**header["jwk"])
        if "d" in header["jwk"]:
            raise ValueError("proof header must not embed a private key")
        jkt = embedded.thumbprint()
        if jkt != expected_jkt:
            raise ValueError("proof key does not match token cnf.jkt")

        signature.verify(embedded)
        payload = json.loads(signature.payload)

        now = time.time()
        iat = payload.get("iat")
        if not isinstance(iat, int):
            raise ValueError("missing iat")
        if iat > now + _PROOF_CLOCK_SKEW_SECONDS:
            raise ValueError("proof issued in the future")
        if now - iat > _PROOF_MAX_AGE_SECONDS:
            raise ValueError("proof too old")

        if payload.get("htm") != htm.upper():
            raise ValueError("htm mismatch")
        if payload.get("htu") != htu:
            raise ValueError("htu mismatch")
        if payload.get("ath") != access_token_hash(access_token):
            raise ValueError("ath mismatch")
        if not isinstance(payload.get("jti"), str):
            raise ValueError("missing jti")

        return VerifiedPopProof(jti=payload["jti"], jkt=jkt)
    except ToolgateError:
        raise
    except Exception as err:
        raise ToolgateError(
            ErrorCodes.PROOF_INVALID, f"proof-of-possession rejected: {err}"
        ) from err
